### ISSUE TITLE
WebGL support?

### ISSUE THREAD

**Issue by skryshi** [mixpanel/mixpanel-unity#10]
Any chance we can get this library to support WebGL sometimes soon?

**Comment by samgreen** [comment]
Hey @skryshi, would it be possible for you to use [Mixpanel JavaScript](https://mixpanel.com/help/reference/javascript) in your game?

Not sure if anything is actually required on our side.

**Comment by skryshi** [comment]
That's what I am doing now, but it isn't terribly convenient. I have to write wrappers for all mixpanel calls like this:

```
public void CustomEvent (string customEventName, mixpanel.Value data) {
#if DISABLE_NATIVE_MIXPANEL
    Application.ExternalCall("ConnectMixpanel.Track", customEventName, data.toStyledString());
#else
    mixpanel.Mixpanel.Track (customEventName, data);
#endif
}
```

I also need a javascript lib to convert `data` in ExternalCall from string to an object before issuing mixpanel calls. If I want to handle callbacks from javascript, then need to write even more wrappers.

It's certainly doable, but far from plug-in-play.

**Comment by samgreen** [comment]
Wow @skryshi thanks for bringing this to our attention. That's certainly not the developer experience that we'd like you to have. We'll discuss internally and figure out what the timeline will be for solid webGL support. For reference, we have tested this library on the following platforms:

- Windows
- Mac
- Android
- iOS

I'm leaving this ticket open until we have an answer for you.

**Comment by samgreen** [comment]
We're investigating webGL support this week.

**Comment by samgreen** [comment]
Looks like we're going to need to do some additional work to get support for JS. Specifically, since we lose support for threads on the web we'll have to re-architect the net workers.

**Comment by ycandrewkao** [comment]
I want to follow up and see if WebGL is something that's supported at this point? Or if not, when it may be? Thx!

**Comment by joecodespark** [comment]
Also interested in WebGL - subscribing to the thread.
@skryshi Did the Javascript solution work out ok?

**Comment by skryshi** [comment]
@joecodespark Yes, Javascript solution worked out ok.

**Comment by rocktavious** [comment]
@zihejia Any word on if MixPanel will support WebGL in Unity?

@skryshi any chance you could point me at an example of your solution?  Mainly i'm confused by `I also need a javascript lib to convert data in ExternalCall from string to an object before issuing mixpanel calls.`

Would really love to see WebGL supported in MixPanel's library.

**Comment by SamLucyk** [comment]
> @zihejia Any word on if MixPanel will support WebGL in Unity?
> 
> @skryshi any chance you could point me at an example of your solution? Mainly i'm confused by `I also need a javascript lib to convert data in ExternalCall from string to an object before issuing mixpanel calls.`
> 
> Would really love to see WebGL supported in MixPanel's library.

I would also love to see an example if anyone has one!

**Comment by rocktavious** [comment]
@SamLucyk I'm in the process of rewriting this library to be C# native (not a c extension) if you interested i could post it up on a github repo for you to check out.  I'm like 85-90% of the way there and have it in a project currently emitting to mixpanel.

**Comment by SamLucyk** [comment]
@rocktavious That would be amazing. That would be a huge help - I would love to check it out if it wouldn't be too much trouble to put it in a github repo. Thanks in advance!

**Comment by rocktavious** [comment]
@SamLucyk https://github.com/rocktavious/mixpanel-unity
I've also set it up so it can be used as a unity package (if you are using > 2018.3)
just modify your `<unity project>/Packages/manifest.json` with 
```
{
  "dependencies": {
    "com.mixpanel.unity": "https://github.com/rocktavious/mixpanel-unity.git",
    ...
  }
}
```
and it will install - otherwise just copy in the Mixpanel folder from the repo to your project.

The API and everything should be 100% compliant with the original API it just no longer is written in C so unity can properly compile the native c# it to any of its platforms (IE WebGL - which is the platform i'm testing it on)

I've also updated the way you configure it so you don't have to have a component in your scene.  Just open `Edit -> Project Settings -> Mixpanel` and you should see roughly the same configuration setup for token and debug token as described in the original docs from mixpanel.

Let me know if you run into any bugs or post them as issues on my github fork - i'm trying to use mixpanel in a project of mine so i'm trying to get it into a working state - as far as i can tell it works and works well, but i may have missed something that someone else might find (it was a massive rewrite and i tried to follow their c code as best i could).  Cheers!

**Comment by SamLucyk** [comment]
@rocktavious Amazing! Thanks so much! I'll let you know any issues I run into and I'll post them on github as well. Trying it out this morning - this will save us a lot of time, Thanks again!

**Comment by SamLucyk** [comment]
@rocktavious Couldn't find a place to add an issue to the repo - let me know if you enable them. But having an issue on a WebGL build, confirmed not having the same issue with Android builds. When trying to access Mixpanel - for example the first Mixpanel.Identify() call - This error logs in the browser console on Chrome.

`MissingMethodException: Method not found: 'Default constructor not found...ctor() of System.ComponentModel.TypeConverter'.
UnityLoader.js:21   at System.Activator.CreateInstance (System.Type type, Boolean nonPublic) [0x00000] in <filename unknown>:0`

Trying to Debug - I'll post back if I find anything. But just wondering if you had any luck on a WebGL build?

**Comment by rocktavious** [comment]
@SamLucyk Sorry didn't realize the issues were turned off - i've enabled them now.  But yes i've successfully built a WebGL build (Unity 2018.3.11)  The error you posted doesn't seem to be referencing any mixpanel code.  I'll try later tonight to setup a fresh project from scratch - maybe if you could give me the flow of the code your using so i can test that exactly that would be helpful.   I'll check over the code and see if i can find anything, but there are no calls to System.Activator.CreateInstance in the library at all.

**Comment by SamLucyk** [comment]
@rocktavious Yeah I agree I can't see why it would only be happening on WebGL - but I've confirmed definitely only happening when I refer to Mixpanel. Created a build with all the Mixpanel References commented out and didn't have the error anymore. It may be something to the way I imported it but I'm just not sure why it works for Android but wouldn't work for WebGL. Again I'll keep looking and message back.

And just to be clear - I was able to successfully build a WebGL build (Unity 2018.3.3) the build just crashes once it gets to a line where Mixpanel is referred to. Maybe I should try updating Unity to 2018.3.11 and try just to rule that out but I'll write back here with anything I find.

**Comment by SamLucyk** [comment]
Confirmed issue is still there building with 2018.3.11 - Also tried importing through the package method and just dragging into the project and same issue is there for both. Going to try an empty project with it now to test some more.

**Comment by rocktavious** [comment]
@SamLucyk ok - very weird - if you can get me a unity project that exhibits the issue i can try debugging on my side too.  I'm running my WebGL builds on Windows 10.  I did a test in an empty project and it emits fine to mixpanel - i did find an issue with the "distinct_id" in WebGL is always "n/a" until you call the identify `Mixpanel.Identify("Hello World")` but making that call works and the future `Mixpanel.Track` calls use the given identity - so i'll fix this bug tonight because it should always have a value (even if you never call `Mixpanel.Identify` and the point of using Identify is to specifically hook in your custom player id info.

Thanks for testing this out - i'm confident we can track it down and get it working for you.  By any chance since the same code works on Android - does the "distinct_id" come through just fine on Android if you call `Mixpanel.Track` prior to calling `Mixpanel.Identify` - there is some untested code in there for getting the native Android identifier - i pulled it directly from mixpanel's original code so it "should" work but i personally have no way to test it.

**Comment by SamLucyk** [comment]
@rocktavious Okay also just confirmed that end empty project emitting a Track works correctly - there must be some conflict with my project. I'm going to try to debug some more and see if I can see what's wrong. I also noticed the distict_id issue -> seems to work from editor but not the WebGL Empty Test Project I made - like you said.

I also noticed an issue with Register on the real project I'm working with even in the editor but was able to fix it with a couple lines change in that function. I'm gonna move this conversation over to the repo and I'll post the issue link here as well.
